include/ginkgo/core/base/mpi.hpp

```diff
@@ -88,18 +102,93 @@ GKO_REGISTER_MPI_TYPE(int, MPI_INT);
 GKO_REGISTER_MPI_TYPE(unsigned short, MPI_UNSIGNED_SHORT);
 GKO_REGISTER_MPI_TYPE(unsigned long, MPI_UNSIGNED_LONG);
 GKO_REGISTER_MPI_TYPE(long, MPI_LONG);
+GKO_REGISTER_MPI_TYPE(long long, MPI_LONG_LONG_INT);
 GKO_REGISTER_MPI_TYPE(float, MPI_FLOAT);
 GKO_REGISTER_MPI_TYPE(double, MPI_DOUBLE);
 GKO_REGISTER_MPI_TYPE(long double, MPI_LONG_DOUBLE);
 GKO_REGISTER_MPI_TYPE(std::complex<float>, MPI_C_COMPLEX);
 GKO_REGISTER_MPI_TYPE(std::complex<double>, MPI_C_DOUBLE_COMPLEX);
 
 
-template <typename T>
-inline const T* in_place()
-{
-    return reinterpret_cast<const T*>(MPI_IN_PLACE);
-}
+/**
+ * A move-only wrapper for a contiguous MPI_Datatype.
+ *
+ * The underlying MPI_Datatype is automatically created and committed when an
+ * object of this type is constructed, and freed when it is destructed.
+ */
+class contiguous_type {
+public:
+    /**
+     * Constructs a wrapper for a contiguous MPI_Datatype.
+     *
+     * @param count  the number of old_type elements the new datatype contains.
+     * @param old_type  the MPI_Datatype that is contained.
+     */
+    contiguous_type(int count, MPI_Datatype old_type) : type_(MPI_DATATYPE_NULL)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Type_contiguous(count, old_type, &type_));
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Type_commit(&type_));
+    }
+
+    /**
+     * Constructs empty wrapper with MPI_DATATYPE_NULL.
+     */
+    contiguous_type() : type_(MPI_DATATYPE_NULL) {}
+
+    /**
+     * Disallow copying of wrapper type.
+     */
+    contiguous_type(const contiguous_type&) = delete;
+
+    /**
+     * Disallow copying of wrapper type.
+     */
+    contiguous_type& operator=(const contiguous_type&) = delete;
+
+    /**
+     * Move constructor, leaves other with MPI_DATATYPE_NULL.
+     *
+     * @param other  the to be moved from object.
```

Comment:
```suggestion
     * @param other  to be moved from object `other`.
```

Reply:
would the following be acceptable?
```suggestion
     * @param other  to be moved from object.
```